bin/run_docker.py

```diff
@@ -0,0 +1,138 @@
+                log_text = log_text.decode("utf-8")
+            log_file.write(log_text.encode("ascii", "ignore").decode("ascii"))
+        else:
+            log_file.write("No Logs")
+
+
+def store_log_file(project, log_filename, submission_id):
+    """Store log file"""
+    statinfo = os.stat(log_filename)
+    if statinfo.st_size > 0:
+        create_folders.create_folders(
+            project_name=project,
+            submission_id=submission_id,
+            create_or_update="update",
+            predictions_file=None,
+            log_file=log_filename,
+        )
+
+
+def store_predictions_file(project, predictions_filename, submission_id):
+    """Store predictions file on Synapse"""
+    statinfo = os.stat(predictions_filename)
+    if statinfo.st_size > 0:
+        create_folders.create_folders(
+            project_name=project,
+            submission_id=submission_id,
+            create_or_update="update",
+            predictions_file=predictions_filename,
+            log_file=None,
+        )
+
+
+def mount_volumes():
+    """
+    A function to mount volumes onto a docker container.
+    """
+    # These are the volumes that you want to mount onto your docker container
+    # output_dir = os.path.join(os.getcwd(), "output")
+    output_dir = os.path.join(os.getcwd(), "output")
+    input_dir = os.path.join(os.getcwd(), "input")
+
+    # These are the locations on the docker that you want your mounted
+    # volumes to be + permissions in docker (ro, rw)
+    # It has to be in this format '/output:rw'
+    mounted_volumes = {output_dir: "/output:rw", input_dir: "/input:ro"}
+    # All mounted volumes here in a list
+    all_volumes = [output_dir, input_dir]
+    # Mount volumes
+    volumes = {}
+    for vol in all_volumes:
+        volumes[vol] = {
+            "bind": mounted_volumes[vol].split(":")[0],
+            "mode": mounted_volumes[vol].split(":")[1],
+        }
+
+    return volumes
+
+
+def run_docker(project, submission_id):
+    """
+    A function to run a Docker container with the specified image and handle any exceptions that may occur.
+    """
+    # Get the Synapse authentication token from the environment variable
+    synapse_auth_token = os.environ.get("SYNAPSE_AUTH_TOKEN")
+
+    # Communication with the Docker client
+    client = docker.from_env()#DockerClient(base_url="unix://var/run/docker.sock")
+
+    # Login to the Docker registry using SYNAPSE_AUTH_TOKEN
+    client.login(username="foo", password=synapse_auth_token, registry="https://docker.synapse.org")
+
+    # Mount the input/ and output/ volumes that will exist in the submission container
+    volumes = mount_volumes()
+
+    # Removing the extra whitespace from the ``docker_image`` string.
+    # TODO: We shouldn't have to do this logic. Figure out what's causing the whitespace
+    # at the end of the docker_image string and remove it.
+    docker_image = get_submission_image.get_submission_image(submission_id)
```

Comment:
I was unable to rely on the output channel of `GET_IMAGE_SUBMISSION.nf` because of some weird tab-ing behavior that was breaking my call to `run_docker`. Here is an [example](https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/1iX4VAqOKALtDn) of a run that broke because of this. You can see under `Command executed` that the image name is tabbed at the end, leaving a whitespace that throws off my docker run in `run_docker.py`.

Here is further evidence - it's the shell command that was executed that led to the error. You can move your cursor through this file and see that there is a random whitespace here: https://s3.console.aws.amazon.com/s3/object/example-dev-project-tower-scratch?region=us-east-1&bucketType=general&prefix=work/0a/17834bca82ca53ff02bbde8989d9cd/.command.sh

I figured the least invasive approach is to reconstruct the submission image within `run_docker.py` and no longer rely on the output tuple of the `GET_SUBMISSION_IMAGE` workflow.

Reply:
I can confirm this change allowed the workflow to run successfully: https://tower-dev.sagebionetworks.org/orgs/Sage-Bionetworks/workspaces/example-dev-project/watch/5uUAWGP3yBEcEr

Reply:
This is interesting. Is there something in `get_submission_image.py` causing this?

Reply:
Nope, I ran `get_submission_image.py` locally and tried to recreate the error, but that script is fine; I think it's something related to how stdouts are passed between tasks. I think at one point the submission ID and the image ID become a tuple, and maybe because of the length of the image_id, the formatting gets weird.

Not a huge deal, but definitely something that would be nice to investigate. I haven't made a tech debt ticket for this yet but plan to, and will update this thread when I do.

Reply:
I think its a semi-big deal if we have to re-route the workflow to get around it... We have used this pattern before and never had this issue. We even use it in other places in this workflow.

Also, in addition to my [comment](https://github.com/Sage-Bionetworks-Workflows/nf-synapse-challenge/pull/12#pullrequestreview-1923593820) about collapsing responsibilities of workflow processes, we are also now creating interconnections between these python scripts which don't have any sort of testing for their logic which isn't great.

Reply:
> We have used this pattern before and never had this issue. We even use it in other places in this workflow.

The issue is that this argument (image_id) is being turned into a string and still keeps the incorrect newline indent. `docker.run` is not forgiving when it comes to the syntax of the `image_id`.

> we are also now creating interconnections between these python scripts which don't have any sort of testing for their logic which isn't great.

What is the difference between the python scripts communicating with each other and the python scripts communicating with the workflows that make one more harmful than the other? Either way I think we should build a test suite for these python scripts. But to address your main point: if we don't want the interconnection between .py's one solution I can propose for this case is to have `run_docker.py` absorb the functionality in `get_submission_image.py`. `get_submission_image.py` only has 1 function which falls into the scope of `run_docker.py`, and I think would fit nicely in there without having to deal with the weird reformatting of the stdout in this case.

Reply:
I agree that `get_submission_image.py` could be absorbed into `run docker.py`. That makes a lot of sense to me. 

To me, the difference is that the processes are executing the scripts, and they are sort of agnostic to what script they are running more so than having Python scripts with strict dependencies on each other. 

Ideally, each process that needs Python is executing the simplest possible Python script to accomplish a clearly defined goal.

Reply:
I see - then for this case, merging the `get_submission_image.py` and `run_docker.py` will provide the least amount of overhead while also sticking to this philosophy for the `bin/` scripts. I think perhaps in April it might be worth doing an architecture revisit, if there is enough time, to refactor and simplify some of these scripts that have grown in scope.

Reply:
Agreed. I think a lot of simplification could be done by adding some of our logic in `bin` into packages like the `synapseclient` (along with tests). For now, I think this is a good solution.

Reply:
I drafted a [small Epic](https://sagebionetworks.jira.com/browse/ORCA-312) for this initiative as I think it will be relevant for PEGS and future challenges. Feel free to modify as you wish.